src/ControlSystem/ControlErrors/SizeControlStateAhSpeed.cpp

```diff
@@ -0,0 +1,112 @@
+    // speed_tolerance_for_char_speed_in_danger is slightly greater than
+    // unity so that we don't panic if the actual char speed is large
+    // enough compared to the target char speed.
+    constexpr double speed_tolerance_for_char_speed_in_danger = 1.1;
+    // We don't want to panic unless crossing time is less than
+    // the current damping time.
+    constexpr double time_tolerance_for_char_speed_in_danger = 0.99;
+    if (delta_radius_is_in_danger) {
+      return false;
+    } else if (crossing_time_info.char_speed_will_hit_zero_first and
+               crossing_time_info.t_char_speed <
+                   info->damping_time *
+                       time_tolerance_for_char_speed_in_danger and
+               min_char_speed < info->target_char_speed *
+                                    speed_tolerance_for_char_speed_in_danger) {
+      return true;
+    } else {
+      return false;
+    }
+  }();
+
+  const bool comoving_decreasing_slower_than_char_speeds =
+      not(crossing_time_info.t_char_speed > 0.0 &&
+          crossing_time_info.t_comoving_char_speed > 0.0 &&
+          update_args.min_comoving_char_speed > 0.0 &&
+          update_args.min_comoving_char_speed /
+                  crossing_time_info.t_comoving_char_speed >
+              update_args.min_char_speed / crossing_time_info.t_char_speed);
+
+  if (char_speed_is_in_danger) {
+    constexpr double min_char_speed_increase_factor = 1.01;
+    if (info->target_char_speed < min_char_speed) {
+      // We are already in state 1, and we are in danger.
+      // But TargetSpeed is less than CharSpeedMin, so we don't want
+      // to continue to push CharSpeedMin downward.  Instead, increase
+      // TargetSpeed to be above CharSpeedMin.
+      // We don't do this if DeltaRexIsInDanger, because for DeltaRex
+      // in danger, we might need to drive CharSpeed lower.
+      info->target_char_speed = min_char_speed * min_char_speed_increase_factor;
+    }
+    info->suggested_time_scale = crossing_time_info.t_char_speed;
+  } else if (delta_radius_is_in_danger) {
+    constexpr double target_speed_decrease_factor = 0.125;
+    constexpr double time_tolerance_for_delta_r =
+        0.25 * time_tolerance_for_delta_r_in_danger;
+    if ((crossing_time_info.t_char_speed > 0.0 and
+         crossing_time_info.t_delta_radius >
+             info->damping_time * time_tolerance_for_delta_r) or
+        update_args.min_comoving_char_speed < 0.0) {
+      info->discontinuous_change_has_occurred = true;
+      info->target_char_speed = min_char_speed * target_speed_decrease_factor;
+      info->suggested_time_scale =
+          std::min(info->damping_time, crossing_time_info.t_delta_radius);
+    } else {
+      info->discontinuous_change_has_occurred = true;
+      info->state = SizeControlLabel::DeltaR;
+      info->suggested_time_scale = crossing_time_info.t_delta_radius;
+      // TODO: PreferState3OverState2 goes here.
```

Comment:
If `State3` is one of those states you'll be adding in a future PR, could you change `State3` and `State2` to more descriptive names? (Same throughout this PR).